### PR TITLE
[WIP] L2 projection accuracy investigation

### PR DESCRIPTION
**Pull request by aprokop**

### CODE REVIEW

**Comment by aprokop** [comment]
The results so far: integrating f(x) = 1 over rectangle

perturbation | L2 -way (abs) | Intrepid Way (abs)
------------- | -------------- | --------- 
0.0 | 6.3e-11 | 7.6e-13
0.1 | 9.1e-04 | 2.0e-13
0.2 | 7.3e-03 | 2.8e-14
0.3 | 2.5e-02 | 5.7e-14
0.4 | 5.8e-02 | 1.1e-13

The comparison of Intrepid to L2 is not fully correct as I'm not sure that Intrepid uses the same integration points (I cheated and did not properly compute Jacobian at the integration points used in L2). But the fact is that L2 does lose accuracy.

**Comment by dalg24** [comment]
ok to test

File: packages/Operators/test/reference_implementation/DTK_ReferenceHexMesh.cpp [hunk 1]
```diff
@@ -208,53 +239,19 @@ void ReferenceHexMesh::buildMesh(
...
+    entity_set->getEntity( node_id, 0, hex_nodes[local_id] );
+                GET_NODE(0, i,   j,   k);
+                GET_NODE(1, i+1, j,   k);
+                GET_NODE(2, i+1, j+1, k);
+                GET_NODE(3, i,   j+1, k);
+                GET_NODE(4, i,   j,   k+1);
+                GET_NODE(5, i+1, j,   k+1);
+                GET_NODE(6, i+1, j+1, k+1);
+                GET_NODE(7, i,   j+1, k+1);
+#undef GET_NODE
```
Comment (dalg24) [inline packages/Operators/test/reference_implementation/DTK_ReferenceHexMesh.cpp]: Nice.

**Comment by aprokop** [comment]
Updated results, now including the new DTK approach

perturbation | DTK old | DTK new | Intrepid
------------ | ------- | ------- | --------
0.0 | 6.324e-11 | 6.324e-11 | 7.674e-13
0.1 | 9.090e-04 | 1.904e-12 | 1.990e-13
0.2 | 7.272e-03 | 6.821e-13 | 2.842e-14
0.3 | 2.454e-02 | 1.478e-12 | 5.684e-14
0.4 | 5.818e-02 | 3.524e-12 | 1.137e-13

Not sure of why DTK and Inrepid differ, but at least the ballpark numbers are much better.

**Comment by dalg24** [comment]
Use `make format-cpp` (see [here](http://datatransferkit.readthedocs.io/en/latest/coding_guidelines.html#clangformat))

**Comment by aprokop** [comment]
@dalg24 I know, but I'm still trying to figure out how to get clang-format 3.9 on my Fedora box (it only has 3.8). My last spack install clang-3.9.0 failed for some reason. If I cannot figure its installation, I'll just start using the dtk docker container.

**Comment by dalg24** [comment]
> My last spack install clang-3.9.0 failed for some reason. If I cannot figure its installation, I'll just start using the dtk docker container.

Yeah the Docker container would do the trick.

**Comment by aprokop** [comment]
> My last spack install clang-3.9.0 failed for some reason.

So, apparently, I have been simply running out of space on my machine. Got rid of useless stuff, and it installed just fine.

**Comment by aprokop** [comment]
So I've added a Jacobian test. It works for linear transformations, but hick ups on some nonlinear ones. In a weird way, too. For some nonlinear transformations 8 out of 9 Jacobian coefficients are correct while one is not. Really weird. Is there something wrong in the ReferenceHex implementation of transformations using Intrepid?

**Comment by aprokop** [comment]
Rebasing the branch.

**Pull request closed without merging**